src/pathfinder/yapf/yapf_ship_regions.cpp

```diff
@@ -0,0 +1,329 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+ /** @file yapf_ship.cpp Implementation of YAPF for water regions, which are used for finding intermediate ship destinations. */
+
+#include "../../stdafx.h"
+#include "../../ship.h"
+
+#include "yapf.hpp"
+#include "yapf_ship_regions.h"
+#include "../water_regions.h"
+
+#include "../../safeguards.h"
+
+constexpr int DIRECT_NEIGHBOR_COST = 100;
+
+/** Yapf Node Key that represents a single patch of interconnected water within a water region. */
+struct CYapfRegionPatchNodeKey {
+	WaterRegionPatchDesc m_water_region_patch;
+
+	static_assert(sizeof(TWaterRegionPatchLabel) == sizeof(byte)); // Important for the hash calculation
+
+	inline void Set(const WaterRegionPatchDesc& water_region_patch)
+	{
+		m_water_region_patch = water_region_patch;
+	}
+
+	inline int CalcHash() const { return m_water_region_patch.label | GetWaterRegionIndex(m_water_region_patch) << 8; }
+	inline bool operator == (const CYapfRegionPatchNodeKey &other) const { return CalcHash() == other.CalcHash(); }
```

Comment:
```suggestion
	inline bool operator==(const CYapfRegionPatchNodeKey &other) const { return CalcHash() == other.CalcHash(); }
```
To be consistent with earlier cases.